File: problems/005-transactional-vfs-quota/solutions/python/main.py
import sys

t = sys.stdin.buffer.read().split()
it = iter(t)
P, N, B, I = map(int, (next(it), next(it), next(it), next(it)))
a = [None] * (P + 1)
used = ino = peak_b = peak_i = sticky = 0
out = []
for _ in range(N):
    op = next(it)
    x = int(next(it))
    err = None
    if op == b"CREATE":
        if a[x] is not None:
            err = "EXISTS"
        elif ino == I:
            err = "INODES"
        else:
            a[x] = 0
            ino += 1
    elif op == b"UNLINK":
        if a[x] is None:
            err = "NOENT"
        else:
            used -= a[x]
            a[x] = None
            ino -= 1
    else:
        if op == b"WRITE":
            off, length = int(next(it)), int(next(it))
            new = (
                a[x]
                if length == 0 and a[x] is not None
                else (0 if length == 0 else max(a[x] or 0, off + length))
            )
        else:
            new = int(next(it))
        if a[x] is None:
            err = "NOENT"
        elif new > a[x] and new - a[x] > B - used:
            err = "BYTES"
        else:
            used += new - a[x]
            a[x] = new
    if err in ("BYTES", "INODES"):
        sticky = 1
    out.append("OK" if err is None else "ERR " + err)
    peak_b = max(peak_b, used)
    peak_i = max(peak_i, ino)
out.append(f"SUMMARY {used} {ino} {peak_b} {peak_i} {sticky}")
print("\n".join(out))

File: problems/005-transactional-vfs-quota/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let p: usize = t.next().unwrap().parse().unwrap();
    let n: usize = t.next().unwrap().parse().unwrap();
    let cap: u64 = t.next().unwrap().parse().unwrap();
    let limit: usize = t.next().unwrap().parse().unwrap();
    let mut ex = vec![false; p + 1];
    let mut sz = vec![0u64; p + 1];
    let (mut used, mut peakb, mut ino, mut peaki, mut sticky) = (0u64, 0u64, 0usize, 0usize, 0);
    let mut out = String::new();
    for _ in 0..n {
        let op = t.next().unwrap();
        let x: usize = t.next().unwrap().parse().unwrap();
        let mut err = "";
        if op == "CREATE" {
            if ex[x] {
                err = "EXISTS"
            } else if ino == limit {
                err = "INODES"
            } else {
                ex[x] = true;
                sz[x] = 0;
                ino += 1
            }
        } else if op == "UNLINK" {
            if !ex[x] {
                err = "NOENT"
            } else {
                used -= sz[x];
                sz[x] = 0;
                ex[x] = false;
                ino -= 1
            }
        } else {
            let v = if op == "WRITE" {
                let off: u64 = t.next().unwrap().parse().unwrap();
                let len: u64 = t.next().unwrap().parse().unwrap();
                if len == 0 {
                    sz[x]
                } else {
                    sz[x].max(off + len)
                }
            } else {
                t.next().unwrap().parse().unwrap()
            };
            if !ex[x] {
                err = "NOENT"
            } else if v > sz[x] && v - sz[x] > cap - used {
                err = "BYTES"
            } else {
                if v >= sz[x] {
                    used += v - sz[x]
                } else {
                    used -= sz[x] - v
                }
                sz[x] = v
            }
        }
        if err.is_empty() {
            out += "OK\n"
        } else {
            out += &format!("ERR {}\n", err);
            if err == "BYTES" || err == "INODES" {
                sticky = 1
            }
        }
        peakb = peakb.max(used);
        peaki = peaki.max(ino);
    }
    out += &format!("SUMMARY {} {} {} {} {}\n", used, ino, peakb, peaki, sticky);
    print!("{}", out);
}
